pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java

```diff
@@ -138,35 +140,34 @@ public Block getNextBlock() {
    */
   private IntermediateResultsBlock combineBlocks()
       throws InterruptedException {
-    final int numOperators = _operators.size();
-    final IntermediateResultsBlock[] blocks = new IntermediateResultsBlock[numOperators];
+    int numOperators = _operators.size();
+    final IntermediateResultsBlock[] intermediateResultsBlocks = new IntermediateResultsBlock[numOperators];
     final CountDownLatch operatorLatch = new CountDownLatch(numOperators);
-
-    final List<AggregationInfo> aggregationsInfo = _brokerRequest.getAggregationsInfo();
-    final int numAggrFunctions = aggregationsInfo.size();
-
-    // TODO: use new aggregation functions to combine results.
-    final List<AggregationFunction> aggregationFunctions =
-        AggregationFunctionFactory.getAggregationFunction(_brokerRequest);
-
     final Map<String, Serializable[]> resultsMap = new ConcurrentHashMap<>();
 
+    List<AggregationInfo> aggregationInfos = _brokerRequest.getAggregationsInfo();
+    final AggregationFunctionContext[] aggregationFunctionContexts =
+        AggregationFunctionUtils.getAggregationFunctionContexts(aggregationInfos, null);
+    final int numAggregationFunctions = aggregationFunctionContexts.length;
+
     for (int i = 0; i < numOperators; i++) {
       final int index = i;
 
       _executorService.execute(new TraceRunnable() {
+        @SuppressWarnings("unchecked")
         @Override
         public void runJob() {
-          AggregationGroupByResult groupByResult;
+          AggregationGroupByResult aggregationGroupByResult;
 
           try {
-            blocks[index] = (IntermediateResultsBlock) _operators.get(index).nextBlock();
-            groupByResult = blocks[index].getAggregationGroupByResult();
+            IntermediateResultsBlock intermediateResultsBlock =
+                (IntermediateResultsBlock) _operators.get(index).nextBlock();
+            intermediateResultsBlocks[index] = intermediateResultsBlock;
+            aggregationGroupByResult = intermediateResultsBlock.getAggregationGroupByResult();
 
-            if (groupByResult != null) {
+            if (aggregationGroupByResult != null) {
```

Comment:
When GroupByOperator hits some exceptions, this will be null. We don't need an else block here because we merge all exceptions anyway.